servers/src/main/java/tachyon/master/file/FileSystemMaster.java

```diff
@@ -877,36 +951,53 @@ private void journalCreatePathResult(InodeTree.CreatePathResult createResult) {
    * @param fileId the source file to rename.
    * @param dstPath the destination path to rename the file to.
    * @return true if the rename was successful
-   * @throws InvalidPathException
-   * @throws FileDoesNotExistException
+   * @throws FileDoesNotExistException if a non-existent file is encountered
+   * @throws InvalidPathException if an invalid path is encountered
+   * @throws IOException if an I/O error occurs
    */
   public boolean rename(long fileId, TachyonURI dstPath)
-      throws InvalidPathException, FileDoesNotExistException {
+      throws FileDoesNotExistException, InvalidPathException, IOException {
     MasterContext.getMasterSource().incRenameOps();
     synchronized (mInodeTree) {
       Inode srcInode = mInodeTree.getInodeById(fileId);
       TachyonURI srcPath = mInodeTree.getPath(srcInode);
+      // Renaming path to itself is a no-op.
       if (srcPath.equals(dstPath)) {
         return true;
       }
+      // Renaming the root is not allowed.
       if (srcPath.isRoot() || dstPath.isRoot()) {
         return false;
       }
+      // Renaming across mount points is not allowed.
+      String srcMount = mMountTable.getMountPoint(srcPath);
+      String dstMount = mMountTable.getMountPoint(dstPath);
+      if ((srcMount == null && dstMount != null)
+          || (srcMount != null && dstMount == null)
+          || (srcMount != null && dstMount != null && !srcMount.equals(dstMount))) {
+        LOG.warn("Renaming " + srcPath + " to " + dstPath + " spans mount points.");
+        return false;
+      }
+      // Renaming onto a mount point is not allowed.
+      if (mMountTable.isMountPoint(dstPath)) {
+        return false;
+      }
+
+      // Rename a path to one of its subpaths is not allowed. Check for that, by making sure
+      // srcComponents isn't a prefix of dstComponents.
       String[] srcComponents = PathUtils.getPathComponents(srcPath.toString());
       String[] dstComponents = PathUtils.getPathComponents(dstPath.toString());
-      // We can't rename a path to one of its subpaths, so we check for that, by making sure
-      // srcComponents isn't a prefix of dstComponents.
       if (srcComponents.length < dstComponents.length) {
         boolean isPrefix = true;
-        for (int prefixInd = 0; prefixInd < srcComponents.length; prefixInd ++) {
+        for (int prefixInd = 0; prefixInd < srcComponents.length; prefixInd++) {
```

Comment:
Minor style, space between `prefixInd` and `++`, also it would be good to uniformly use `Ind` or `Idx` or `i` or some other variant